sstables/compaction_manager.cc

```diff
@@ -731,15 +731,17 @@ future<> compaction_manager::perform_cleanup(column_family* cf) {
         return make_exception_future<>(std::runtime_error(format("cleanup request failed: there is an ongoing cleanup on {}.{}",
             cf->schema()->ks_name(), cf->schema()->cf_name())));
     }
-    return rewrite_sstables(cf, sstables::compaction_options::make_cleanup(), [this] (const table& table) {
-        auto schema = table.schema();
+    return seastar::async([this, cf] {
+        auto schema = cf->schema();
         auto owned_ranges = service::get_local_storage_service().get_local_ranges(schema->ks_name());
         auto sstables = std::vector<sstables::shared_sstable>{};
-        const auto candidates = table.candidates_for_compaction();
+        const auto candidates = cf->candidates_for_compaction();
         std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(sstables), [&owned_ranges, schema] (const sstables::shared_sstable& sst) {
+            seastar::thread::maybe_yield();
             return owned_ranges.empty() || needs_cleanup(sst, owned_ranges, schema);
         });
-        return sstables;
+        rewrite_sstables(cf, sstables::compaction_options::make_cleanup(),
+                [sstables = std::move(sstables)] (const table&) { return sstables; }).get();
```

Comment:
While reviewing ur work, I realized we need to be careful about lambda storing a copy of sstables, because we want rewrite_sstables() to release space of a sstable already rewritten. But it turns out that the callback passed to rewrite_sstables() will be destroyed by the time it returns a future.